source/agora/consensus/protocol/Nominator.d

```diff
@@ -520,22 +517,16 @@ extern(D):
         assert(type >= TimerType.min && type <= TimerType.max);
         if (callback is null || timeout == 0)
         {
-            // signal deactivation of all current timers with this timer type
-            this.active_timer_ids[type] = this.last_timer_id[type] + 1;
+            if (auto timer = this.active_timers[type])
+            {
+                timer.stop();
+                this.active_timers[type] = null;
+            }
             return;
         }
 
-        const timer_id = ++this.last_timer_id[type];
-        this.taskman.runTask(
-        {
-            this.taskman.wait(timeout.msecs);
-
-            // timer was cancelled
-            if (timer_id < this.active_timer_ids[type])
-                return;
-
-            callCPPDelegate(callback);
-        });
+        this.active_timers[type] = this.taskman.setTimer(
+            timeout.msecs, { callCPPDelegate(callback); });
```

Comment:
Looks like previously set timers aren't stopped ?

Reply:
@TrustHenry could you also check if `this.active_timers[type]` exists here, and if it does call stop on it?

Reply:
In the unit test, Timer.stops confirmed a total of 106 occurrences.
This is also if (auto timer = this.active_timers[type])

Reply:
I meant that we should probably do something like this:

```diff
diff --git a/source/agora/consensus/protocol/Nominator.d b/source/agora/consensus/protocol/Nominator.d
index 4a0f115e2..381c6faca 100644
--- a/source/agora/consensus/protocol/Nominator.d
+++ b/source/agora/consensus/protocol/Nominator.d
@@ -515,16 +515,15 @@ extern(D):

         const type = cast(TimerType) timer_type;
         assert(type >= TimerType.min && type <= TimerType.max);
-        if (callback is null || timeout == 0)
+        if (auto timer = this.active_timers[type])
         {
-            if (auto timer = this.active_timers[type])
-            {
-                timer.stop();
-                this.active_timers[type] = null;
-            }
-            return;
+            timer.stop();
+            this.active_timers[type] = null;
         }

+        if (callback is null || timeout == 0)
+            return;
+
         this.active_timers[type] = this.taskman.setTimer(
             timeout.msecs, { callCPPDelegate(callback); });
     }
```

So always stop any existing timers. And then if callback is null we just return, otherwise we set a timer.

Reply:
Was there a leak here?
The Code of @AndrejMitrovic  applied successfully.

Reply:
If this callback is not null?

Reply:
It's possible.